add support for compressibility of log file

miaoyq opened the pull request:
Signed-off-by: Yanqiang Miao <miao.yanqiang@zte.com.cn>

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/docker/docker/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
This PR adds support for compressibility of log file. I added a new option `conpression` for the `jsonfile` log driver, this option allows the user to specify compression algorithm to compress the log files.

**- How I did it**
The `jsonfile` will compress the history log file, except for `container-id-json.log.1`, when `compression=gzip`. The `container-id-json.log.1` is not compressed in order to prevent the log tracking tool from losing some historical log data when a new log file is created.
**- How to verify it**
I run a container as follow:
 ```
$ docker run -d --log-opt max-size=1m --log-opt max-file=5 --log-opt compression=gzip busybox /bin/sh -c 'while true; do sleep 0.001; echo "hello!"; done'
b355968fc501942525f8c48a52201e62cc9e48b6b342b4372d60000feb59c447
```
The results are as follows:
```
$ ls -l
total 1360
-r--------    1 root     root        120003 Jan  5 10:48 b355968fc501942525f8c48a52201e62cc9e48b6b342b4372d60000feb59c447-json.log
-r--------    1 root     root       1000039 Jan  5 10:48 b355968fc501942525f8c48a52201e62cc9e48b6b342b4372d60000feb59c447-json.log.1
-rw-------    1 root     root         75813 Jan  5 10:48 b355968fc501942525f8c48a52201e62cc9e48b6b342b4372d60000feb59c447-json.log.2.gz
-rw-------    1 root     root         75795 Jan  5 10:48 b355968fc501942525f8c48a52201e62cc9e48b6b342b4372d60000feb59c447-json.log.3.gz
-rw-------    1 root     root         75814 Jan  5 10:48 b355968fc501942525f8c48a52201e62cc9e48b6b342b4372d60000feb59c447-json.log.4.gz
drwx------    2 root     root             6 Jan  5 10:47 checkpoints
-rw-r--r--    1 root     root          2532 Jan  5 10:47 config.v2.json
-rw-r--r--    1 root     root          1165 Jan  5 10:47 hostconfig.json
-rw-r--r--    1 root     root            13 Jan  5 10:47 hostname
-rw-r--r--    1 root     root           174 Jan  5 10:47 hosts
-rw-r--r--    1 root     root            75 Jan  5 10:47 resolv.conf
-rw-r--r--    1 root     root            71 Jan  5 10:47 resolv.conf.hash
drwxrwxrwt    2 root     root            40 Jan  5 10:47 shm
```
**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

AkihiroSuda commented on daemon/logger/loggerutils/rotatefilewriter.go:
```diff
@@ -15,11 +19,12 @@ type RotateFileWriter struct {
 	capacity     int64 //maximum size of each file
 	currentSize  int64 // current size of the latest file
 	maxFiles     int   //maximum number of files
+	compress     bool  // whether old versions of log files are compressed
```
I wonder this should be a string (e.g. "gzip") for extensibility

AkihiroSuda commented on daemon/logger/loggerutils/rotatefilewriter.go:
```diff
@@ -98,6 +118,30 @@ func rotate(name string, maxFiles int) error {
 	return nil
 }
 
+func copyToFile(outfile string, r io.Reader) error {
```
Let's use AtomicFileWriter

https://github.com/docker/docker/blob/master/pkg/ioutils/fswriters.go

miaoyq replied:
@AkihiroSuda  The data source is an `io.Reader` interface type, so I think using `io.Copy` might be a bit more appropriate here.

AkihiroSuda commented on daemon/logger/loggerutils/rotatefilewriter.go:
```diff
@@ -80,13 +86,27 @@ func (w *RotateFileWriter) checkCapacityAndRotate() error {
 	}
+
+	suffix := ""
+	if maxFiles > 2 && compress {
```
Please add a comment, UT, and doc?

AkihiroSuda commented:
Design SGTM

AkihiroSuda commented on daemon/logger/jsonfilelog/jsonfilelog.go:
```diff
@@ -119,6 +128,7 @@ func ValidateLogOpt(cfg map[string]string) error {
 		switch key {
 		case "max-file":
 		case "max-size":
+		case "compress":
```
compress -> compression?

miaoyq commented:
Thanks @AkihiroSuda for review, I will refactor the code according to your comments.

thaJeztah commented:
In general, the actual json-log log files are designed for internal use by docker, and not for external consumption. The rotated files are actually used when running `docker logs` (i.e., the reader loops over those files as well). I don't think we should implement this, if it's only to save space, as it (probably) makes these files no longer useful for `docker logs`, and using a different logging driver to collect logs centrally is a better solution for this.

TL;DR, I'm not sure we should implement this

/cc @cpuguy83

cpuguy83 commented:
I don't really see it as harmful. We may just want to compress by default and not even provide a configuration for it.
We can read from compressed files easily enough.
This way docker can store more logs in significantly less space.

miaoyq commented:
@thaJeztah @cpuguy83 @AkihiroSuda , Thank so much for your advice and guidance. 
I have finished this work, while the compressed files are also available for `docker logs`
Here the option `compresstion=gzip` still need to be set, instead of default

My test as follow:
step 1: Create a container with `compression=gzip`
```
$ docker run -d --log-opt max-size=1m --log-opt max-file=5 --log-opt compression=gzip busybox /bin/sh -c 'while true; do sleep 0.001; echo "$(date)"; done'
```
step 2: Check the log files list after about one minute
```
[node1] (local) root@10.0.11.3 /graph/containers/5e597048f230ed657baa51cb70d4ea087257e402f3a0340d480dfbb840514d62
$ ls -l
total 1324
-r--------    1 root     root        142481 Jan 16 03:41 5e597048f230ed657baa51cb70d4ea087257e402f3a0340d480dfbb840514d62-json.log
-r--------    1 root     root       1000072 Jan 16 03:41 5e597048f230ed657baa51cb70d4ea087257e402f3a0340d480dfbb840514d62-json.log.1
-rw-r-----    1 root     root         60488 Jan 16 03:41 5e597048f230ed657baa51cb70d4ea087257e402f3a0340d480dfbb840514d62-json.log.2.gz
-rw-r-----    1 root     root         60134 Jan 16 03:40 5e597048f230ed657baa51cb70d4ea087257e402f3a0340d480dfbb840514d62-json.log.3.gz
-rw-r-----    1 root     root         60479 Jan 16 03:40 5e597048f230ed657baa51cb70d4ea087257e402f3a0340d480dfbb840514d62-json.log.4.gz
drwx------    2 root     root             6 Jan 16 03:38 checkpoints
-rw-r--r--    1 root     root          2534 Jan 16 03:41 config.v2.json
-rw-r--r--    1 root     root          1168 Jan 16 03:41 hostconfig.json
-rw-r--r--    1 root     root            13 Jan 16 03:38 hostname
-rw-r--r--    1 root     root           174 Jan 16 03:38 hosts
-rw-r--r--    1 root     root            75 Jan 16 03:38 resolv.conf
-rw-r--r--    1 root     root            71 Jan 16 03:38 resolv.conf.hash
drwx------    2 root     root             6 Jan 16 03:38 shm
```
step 3: Extract the first timestamp from the oldest log file (`2017-01-16T03:40:09.719852846Z`)
```
$ vi 5e597048f230ed657baa51cb70d4ea087257e402f3a0340d480dfbb840514d62-json.log.4.gz
    1 {"log":"Mon Jan 16 03:40:09 UTC 2017\n","stream":"stdout","time":"2017-01-16T03:40:09.719852846Z"}
    2 {"log":"Mon Jan 16 03:40:09 UTC 2017\n","stream":"stdout","time":"2017-01-16T03:40:09.73350815Z"}
    3 {"log":"Mon Jan 16 03:40:09 UTC 2017\n","stream":"stdout","time":"2017-01-16T03:40:09.733516097Z"}
    4 {"log":"Mon Jan 16 03:40:09 UTC 2017\n","stream":"stdout","time":"2017-01-16T03:40:09.733520285Z"}
    5 {"log":"Mon Jan 16 03:40:09 UTC 2017\n","stream":"stdout","time":"2017-01-16T03:40:09.733524048Z"}
	   ... ...
```
step 4: run `docker logs`
```
$ docker logs -t -f --since "2017-01-16T03:40:09.719852846Z" 5e >> test.log
```
step 5: Interrupt the `docker logs` process and check the contents of `test.log`. We can find all the logs since `2017-01-16T03:40:09.719852846Z`
```
vi test.log
    1 2017-01-16T03:40:09.719852846Z Mon Jan 16 03:40:09 UTC 2017
    2 2017-01-16T03:40:09.733508150Z Mon Jan 16 03:40:09 UTC 2017
    3 2017-01-16T03:40:09.733516097Z Mon Jan 16 03:40:09 UTC 2017
	  ... ...
	  ... ...
41889 2017-01-16T03:41:18.444189279Z Mon Jan 16 03:41:18 UTC 2017
41890 2017-01-16T03:41:18.445763848Z Mon Jan 16 03:41:18 UTC 2017
41891 2017-01-16T03:41:18.447345408Z Mon Jan 16 03:41:18 UTC 2017
41892 2017-01-16T03:41:18.448965201Z Mon Jan 16 03:41:18 UTC 2017
```

LK4D4 commented:
The design looks good. Let's review.
ping @cpuguy83 @AkihiroSuda

LK4D4 commented on daemon/logger/jsonfilelog/read.go:
```diff
@@ -53,19 +55,47 @@ func (l *JSONFileLogger) readLogs(logWatcher *logger.LogWatcher, config logger.R
+			}
+			defer rc.Close()
+
+			fileData, err := ioutil.ReadAll(rc)
```
This kinda looks dangerous. You might read too much into memory.

miaoyq replied:
We can replace this with a temporary file, but it may reduce efficiency.

cpuguy83 replied:
I think we have to, or implement a seeker for a compressed stream.

miaoyq replied:
@LK4D4 @cpuguy83 I have implemented a seeker for a compressed stream. PTAL, 
If there is anything wrong, please point out the right direction, thanks so much!

LK4D4 commented on daemon/logger/loggerutils/rotatefilewriter.go:
```diff
@@ -80,24 +85,86 @@ func (w *RotateFileWriter) checkCapacityAndRotate() error {
+	compressWriter, err := archive.CompressStream(outFile, compression)
+	defer compressWriter.Close()
+
+	fileData, err := ioutil.ReadFile(fileName)
```
this should be `io.Copy` instead of read everything in memory

cpuguy83 replied:
I don't think we want to read this in here. This should stream to the compressed writer.

miaoyq replied:
@LK4D4 updated. PTAL

cpuguy83 commented on daemon/logger/loggerutils/rotatefilewriter.go:
```diff
@@ -80,24 +85,86 @@ func (w *RotateFileWriter) checkCapacityAndRotate() error {
+	extension := ""
+	var compressionAlg archive.Compression
+	if compress != "" {
+		switch compress {
```
Do we need to support a bunch of formats?

cpuguy83 replied:
Not relying on the archive package would definitely be a good thing.

miaoyq replied:
I agree with you, but @AkihiroSuda has different opinions: 
> I wonder this should be a string (e.g. "gzip") for extensibility

cpuguy83 commented on daemon/logger/jsonfilelog/jsonfilelog.go:
```diff
@@ -119,6 +120,7 @@ func ValidateLogOpt(cfg map[string]string) error {
 		switch key {
 		case "max-file":
 		case "max-size":
+		case "compression":
```
I would like to not have a flag and just compress... do you see why people would not want this?

I guess there's people violating the rule of everything in `/var/lib/docker` should be private to the engine.

cpuguy83 replied:
After discussing with others, we decided to not compress by default in this driver as it could break people who aren't expecting it.

LK4D4 commented on daemon/logger/jsonfilelog/jsonfilelog.go:
```diff
@@ -119,6 +120,7 @@ func ValidateLogOpt(cfg map[string]string) error {
 		switch key {
 		case "max-file":
 		case "max-size":
+		case "compression":
```
But they should be punished.

cpuguy83 replied:
10 lashings each.

miaoyq replied:
Do we compress the log file by default? @cpuguy83 @LK4D4

mlaventure commented:
ping @AkihiroSuda @cpuguy83 @LK4D4 looks like @miaoyq is waiting on a consensus regarding the flag (if any)/supported formats

AkihiroSuda commented:
I'm OK for supporting only single compression algorithm at the moment.
We can add algorithms later on demand (if any)

miaoyq commented:
@cpuguy83 @LK4D4 Thanks for your reviews,  I have refactored the code according to your comments.,  PTAL  : )

mlaventure commented:
ping @cpuguy83

AkihiroSuda commented:
@cpuguy83 Design LGTY?

miaoyq commented:
ping @cpuguy83  Could you please spend a little time to review this PR?  Thanks so much!

cpuguy83 commented on pkg/archive/archive.go:
```diff
@@ -204,6 +204,35 @@ func DecompressStream(archive io.Reader) (io.ReadCloser, error) {
+
+// Seek sets the offset for the next Read, and can only seek to the
+// beginning of the file.
+func (g *GzipReadSeekerWrapper) Seek(offset int64, whence int) (int64, error) {
```
I don't think this will work. The seeker will need to do more than seek to `0,0`
We'll probably have to decompress to a temp file to make this work.

miaoyq replied:
@cpuguy83  I have updated the codes, PTAL, Tks  :)

miaoyq commented:
ping @cpuguy83

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "container-log-add-archive" git@github.com:miaoyq/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842354093160
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

cpuguy83 commented on daemon/logger/jsonfilelog/read.go:
```diff
@@ -53,19 +54,57 @@ func (l *JSONFileLogger) readLogs(logWatcher *logger.LogWatcher, config logger.R
 				break
 			}
+
+			cf, err := os.Open(fmt.Sprintf("%s.%d.gz", pth, i-1))
```
Getting the uncompressed stream should be handled outside of this function.

cpuguy83 reviewed (changes requested):
Sorry for the slow review time here, hopefully we can get this in soon.
Some things I'd like to see:

1. Only use gzip, ditch the `archive` package and go straight to the golang built-in `gzip` package.
2. Make compression non-optional
3. Add metadata to the gzip header (thinking at least store the first/last log entry timestamps for optimizing the reader.
i. This can go in the `Extra` field in https://golang.org/src/compress/gzip/gunzip.go?s=1297:1500#L42 with json encoding.
4. Make the reader take advantage of the metadata.

thaJeztah commented:
@cpuguy83 re;

> Make compression non-optional

I know there's software (IIRC, k8s currently) that - although they shouldn't - directly consume the son-file log files; would that be an issue?

miaoyq commented:
@cpuguy83  Thank you very much for giving such valuable comments . :+1: 
I have updated, PTAL

cpuguy83 commented on daemon/logger/loggerutils/rotatefilewriter.go:
```diff
@@ -90,20 +96,125 @@ func rotate(name string, maxFiles int) error {
+	return nil
+}
+
+func readLastLine(file *os.File) ([]byte, error) {
```
We may want to change the interface on the rotating writer to be `Write(logEntry) error` instead of a generic io.Writer.

miaoyq replied:
>Maybe we can keep the last timestamp in memory somehow... This just being a generic writer I'm not sure right now the best approach.
>
>:( This may be getting too much in the way of the log writer in the name of optimizing reads... will have to think on this. Do you have any ideas?

@cpuguy83  I originally thought we should not compress the `container-id.log.1` log file so that other log tracking tools do not lose any history log when a new log file is created. In this case, we have to read the `last timestamp` from the file, because the data in the memory will be lost when the `dockerd ` is restarted.

However, if `jsonfilelog` is only used for the `docker log`, then lasttime is stored in memory is the better solution. :+1:

miaoyq replied:
> We may want to change the interface on the rotating writer to be Write(logEntry) error instead of a generic io.Writer.

@cpuguy83  Updated.

thaJeztah commented:
@miaoyq looks like this needs a rebase :cry:

miaoyq commented:
@thaJeztah  Done, Thanks for reminding! :smile:

cpuguy83 commented on daemon/logger/loggerutils/rotatefilewriter.go:
```diff
@@ -90,20 +96,125 @@ func rotate(name string, maxFiles int) error {
+			return err
+		}
+
+		if err := compressJSONFile(name + ".2"); err != nil {
```
It would be nice if we can handle compression asynchronously from log writes.
We'd probably have to have a write lock and a rotation lock, basically the next rotation would be blocked until compression is completed but writing can continue up until it needs to rotate again.

miaoyq replied:
@cpuguy83  Updated, PTAL, Thanks so much.

cpuguy83 commented on daemon/logger/loggerutils/rotatefilewriter.go:
```diff
@@ -33,45 +44,70 @@ func NewRotateFileWriter(logPath string, capacity int64, maxFiles int) (*RotateF
+		w.writerMu.Unlock()
+		return errors.New("cannot write because the output file was closed")
 	}
+	w.writerMu.Unlock()
```
I don't think it's safe to unlock here

cpuguy83 commented on daemon/logger/loggerutils/rotatefilewriter.go:
```diff
@@ -33,45 +44,70 @@ func NewRotateFileWriter(logPath string, capacity int64, maxFiles int) (*RotateF
 
-	n, err := w.f.Write(message)
+	w.writerMu.Lock()
+	defer w.writerMu.Unlock()
```
We don't want to use `defer` in this function since this is in a hot path and `defer` carries with it a significant amount of overhead.

cpuguy83 commented on daemon/logger/loggerutils/rotatefilewriter.go:
```diff
@@ -33,45 +44,70 @@ func NewRotateFileWriter(logPath string, capacity int64, maxFiles int) (*RotateF
+		defer func() {
+			w.rotateMu.Unlock()
+			if err == nil && w.maxFiles > 1 {
+				<-ch
```
This is essentially the same as running the compressor in this goroutine, which we don't need to do.
I think what we can do here is something like:


```go
w.rotateMu.Lock()
// stuff
w.rotate(name)
// other stuff

if notCompressing { // whatever this condition needs to be
    w.rotateMu.Unlock()
    return nil
}
go func() {
    compress(name)
    w.rotateMu.Unlock()
}()
return nil
```

This way compression does not block writes but does block rotates.

cpuguy83 replied:
And I don't think the compressor will need to be bound to `RotateFileWriter`

miaoyq replied:
@cpuguy83  Indeed this way is more concise and clear, Thanks!

cpuguy83 commented on daemon/logger/loggerutils/rotatefilewriter.go:
```diff
@@ -33,45 +44,70 @@ func NewRotateFileWriter(logPath string, capacity int64, maxFiles int) (*RotateF
-func (w *RotateFileWriter) Write(message []byte) (int, error) {
-	w.mu.Lock()
+//WriteLog write log logEntry to File
+func (w *RotateFileWriter) Write(logEntry *jsonlog.JSONLogs) error {
```
This would be nice to decouple from jsonlogs but doesn't have to be now.

cpuguy83 replied:
What I was thinking here is we instantiate `RoteFileWriter` with a `LogEntryEncoder`, which would be a new type like:

```go
type LogEntryEncoder interface {
    Encode(l LogEntry) error
}
```

And the jsonlog driver would provide an encoder that can convert the `LogEntry` into the jsonlogs encoded format.... which is basically everything that the json log driver's `Log()` method does, but encapsulated into a separate type.

And then change the name of this function to `WriteLog` or `WriteLogEntry` so it's not ever confused with an `io.Writer`.

miaoyq replied:
@cpuguy83 
It's really nice, I will implement it in a new PR, Thanks for your guidance!  : )

cpuguy83 reviewed (changes requested):
Looking good.
A couple of changes.

We should probably also move the log reading to the `RotateFileWriter` object

miaoyq commented:
> We should probably also move the log reading to the `RotateFileWriter` object
I have updated, but the above sentence is not quite understand, trouble to re-explain please, thanks so much!

AkihiroSuda commented on daemon/logger/loggerutils/rotatefilewriter.go:
```diff
@@ -1,23 +1,34 @@
 	"github.com/docker/docker/pkg/pubsub"
 )
 
+const defaultTimestamp = "0001-01-01T00:00:00Z"
```
Why do we need this default time stamp?

miaoyq replied:
@AkihiroSuda  Make sure that the `lastTimestamp` is a `time` format string.

AkihiroSuda replied:
If you don't prefer an empty string here, 1970-01-01T00:00:00Z would be consistent with most other programs.

PTAL @cpuguy83 WDYT?

cpuguy83 replied:
I think an empty string would be fine here.

miaoyq replied:
@AkihiroSuda @cpuguy83  Updated.

thaJeztah commented:
ping @cpuguy83 ^^ https://github.com/moby/moby/pull/29932#discussion_r120261013

cpuguy83 commented on daemon/logger/jsonfilelog/read.go:
```diff
@@ -111,6 +144,51 @@ func (l *JSONFileLogger) readLogs(logWatcher *logger.LogWatcher, config logger.R
+
+	// Extract the last log entry timestramp from the gzip header
+	extra := &loggerutils.RotateFileMetadata{}
+	err = json.Unmarshal(rc.Header.Extra, &extra)
```
this is a double pointer

cpuguy83 commented on daemon/logger/jsonfilelog/read.go:
```diff
@@ -111,6 +144,51 @@ func (l *JSONFileLogger) readLogs(logWatcher *logger.LogWatcher, config logger.R
+	}
+	defer rs.Close()
+
+	_, err = io.Copy(rs, rc)
```
Can we use `pools.Copy` for this from `github.com/pkg/pools`?

cpuguy83 commented on daemon/logger/loggerutils/rotatefilewriter.go:
```diff
@@ -101,13 +142,62 @@ func rotate(name string, maxFiles int) error {
+	}
+
+	file.Seek(0, os.SEEK_SET)
+	_, err = io.Copy(compressWriter, file)
```
`pools.Copy`

cpuguy83 commented on daemon/logger/loggerutils/rotatefilewriter.go:
```diff
@@ -101,13 +142,62 @@ func rotate(name string, maxFiles int) error {
 func (w *RotateFileWriter) LogPath() string {
-	w.mu.Lock()
-	defer w.mu.Unlock()
+	w.writerMu.Lock()
```
I don't think we need the writer mutex here.

miaoyq replied:
@cpuguy83  Yes, `rotateMu` is enough, Thanks  : )

cpuguy83 commented on daemon/logger/loggerutils/rotatefilewriter.go:
```diff
@@ -1,23 +1,32 @@
+	f             *os.File // store for closing
+	closed        bool
+	buf           *bytes.Buffer
+	writerMu      sync.Mutex
```
Can you add comments for each mutex explaining when/where they are used.

miaoyq commented:
@cpuguy83  Updated, PTAL,  Thanks! : )

AkihiroSuda commented:
@cpuguy83 PTAL?

AkihiroSuda commented:
ping @cpuguy83

cpuguy83 commented:
Right now I'm worried about some of the complexity here. Also what happens if we try to read before a rotated file is compressed?

Taking the metadata idea a bit further, maybe we can store metadata in bolt and move much of the read logic dealing with files and such into rotate object.

miaoyq commented:
Thanks @cpuguy83.
Can we use a file lock to solve the problem?  
The rotated file will be remove only after it is compressed completely,  so we can lock the rotated file when we try to read, and the remove operation will be block until the file lock is released.

cpuguy83 commented:
@miaoyq Not a big fan of the file lock strategy, however moving read logic into the rotater would allow us to re-use the rotate lock during reads.

cpuguy83 commented:
@miaoyq I've moved around the read logic in #34162

miaoyq commented:
@cpuguy83  I’m so sorry for the delay in my reply.
It is a good opinion that move the read logic into the rotater. I will refactor the code based on #34162.

AkihiroSuda commented:
ping @miaoyq

thaJeztah commented:
Looks like this depends on https://github.com/moby/moby/pull/34162, which isn't merged yet

miaoyq commented:
@AkihiroSuda This PR depend on #34162, 
 @thaJeztah Yes. : )

miaoyq commented:
ping @thaJeztah @AkihiroSuda @cpuguy83 
 I have refactored the code based on #34162, PTAL. Tks.

thaJeztah commented:
Thanks for updating! @cpuguy83 PTAL

cpuguy83 commented on daemon/logger/jsonfilelog/jsonfilelog.go:
```diff
@@ -61,6 +61,15 @@ func New(info logger.Info) (logger.Logger, error) {
 	}
 
+	var compress = true
+	if compressString, ok := info.Config["compression"]; ok {
```
Do we need to validate that this is only set when log rotation is enabled?

miaoyq replied:
OK, will do.

cpuguy83 commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -2,32 +2,47 @@ package loggerutils
+
+// RotateFileMetadata is a metadata of the gzip header of the compressed log file
+type RotateFileMetadata struct {
+	LastTime string `json:"lastTime,omitempty"`
```
Can we use `time.Time` here?

cpuguy83 replied:
This could also likely do well to have a first and last timestamp.

miaoyq replied:
>Can we use time.Time here?

Will do.
> This could also likely do well to have a first and last timestamp.
In general, the last timestamp of previous log file is the first timestramp the current log file, so I think last timestamp is enough.

cpuguy83 commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -2,32 +2,47 @@ package loggerutils
+	currentSize   int64        // current size of the latest file
+	maxFiles      int          //maximum number of files
+	compress      bool         // whether old versions of log files are compressed
+	lastTimestamp string       // timestamp of the last log
```
time.Time

miaoyq replied:
Will do.

cpuguy83 commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -116,24 +147,72 @@ func rotate(name string, maxFiles int) error {
+	defer compressWriter.Close()
+
+	// Add the last log entry timestramp to the gzip header
+	extra := &RotateFileMetadata{}
```
No need for pointer?

miaoyq replied:
Will do.

miaoyq commented:
@cpuguy83  Thanks for review. :-)

miaoyq commented:
@cpuguy83  Addressed all comments except [this](https://github.com/moby/moby/pull/29932#discussion_r149200621), PTAL.

thaJeztah commented:
@miaoyq did you close by accident?

miaoyq commented:
@thaJeztah  Yes,  have reopened, :sweat_smile:

miaoyq commented:
@cpuguy83 @thaJeztah  Will address the comments. :)

miaoyq commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -158,29 +243,32 @@ func (w *LogFile) Close() error {
 		for _, f := range files {
 			f.Close()
+			if strings.HasSuffix(f.Name(), tmpLogfileSuffix) {
+				os.Remove(f.Name())
```
@cpuguy83 Done.
I think we could stay this way before finding a better solution. : )

miaoyq commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -144,8 +227,10 @@ func (w *LogFile) MaxFiles() int {
-	defer w.mu.Unlock()
+	w.writerMu.Lock()
+	defer w.writerMu.Unlock()
+	w.rorateMu.Lock()
```
Done.

miaoyq commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -93,47 +110,113 @@ func (w *LogFile) checkCapacityAndRotate() error {
 func (w *LogFile) LogPath() string {
-	w.mu.Lock()
-	defer w.mu.Unlock()
+	w.writerMu.Lock()
```
Done.

miaoyq commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -93,47 +110,113 @@ func (w *LogFile) checkCapacityAndRotate() error {
 }
 
+func compressJSONFile(fileName string, lastTimestamp time.Time) {
+	var err error
```
Done.

miaoyq commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -2,32 +2,46 @@ package loggerutils
-	capacity      int64 //maximum size of each file
-	currentSize   int64 // current size of the latest file
-	maxFiles      int   //maximum number of files
+	writerMu      sync.RWMutex // protects the logfile access
```
Done.

miaoyq commented:
@cpuguy83 @thaJeztah  Thanks for review, have addressed all comments. :)

miaoyq commented:
/rebuild/janky

miaoyq commented:
ping @cpuguy83

miaoyq commented:
@cpuguy83  Could you please spend a little time in reviewing this pr? Thanks so much. :)

cpuguy83 commented on daemon/logger/jsonfilelog/jsonfilelog.go:
```diff
@@ -61,6 +61,17 @@ func New(info logger.Info) (logger.Logger, error) {
 		}
 	}
 
+	var compress = false
```
nit `var compress bool` is enough

miaoyq replied:
Will do.

cpuguy83 commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -232,6 +342,46 @@ func (w *LogFile) openRotatedFiles() (files []*os.File, err error) {
+	if err != nil {
+		return "", err
+	}
+	defer rs.Close()
```
How come we close here instead of returning the open file?

miaoyq replied:
will do.

miaoyq replied:
Done.

cpuguy83 commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -224,6 +312,28 @@ func (w *LogFile) openRotatedFiles() (files []*os.File, err error) {
 				return nil, err
 			}
+
+			tmpFile, err := decompressedJsonfile(fmt.Sprintf("%s.%d.gz", w.f.Name(), i-1), config.Since)
```
This needs to be able to handle parallel readers. I don't think it will right now.
Maybe we can keep an index of decompressed files that are ref-counted and removed at 0 references?

cpuguy83 replied:
We may also want to optimize how we are opening files (and as such decompressing them) with regard to the number of lines requested with tail.

miaoyq replied:
>A good follow-up to this PR would be to to copy the file data in kernel space where available: https://github.com/moby/moby/blob/master/daemon/graphdriver/copy/copy.go

I think we could open the same file multiple times instead of copying the file (even hard links), only record the corresponding references-counted, looks like:
```go
    decompressedFileName := fmt.Sprintf("%s.%d.gz", w.f.Name(), i-1) + tmpLogfileSuffix
    var tmpFile *os.File
    // w.filesRefCounter  map[string]int
    if _, ok := w.filesRefCounter[decompressedFileName]; ok {
            w.filesRefCounter[decompressedFileName]++
            tmpFile, err = os.Open(decompressedFileName)
            if err != nil {
                    if !os.IsNotExist(err) {
                            return nil, err
                    }
                    continue
            }
    } else {
            tmpFile, err = decompressedJsonfile(fmt.Sprintf("%s.%d.gz", w.f.Name(), i-1), config.Since)
            if err != nil {
                    if !os.IsNotExist(err) {
                            return nil, err
                    }
                    continue
            }
            if tmpFile == nil {
                    // The log before `config.Since` does not need to read
                    break
            }
            w.filesRefCounter[tmpFile.Name()] = 1
    }

    files = append(files, tmpFile)
```
@cpuguy83 wdyt?

cpuguy83 replied:
More or less +1.
- The ref counter needs to be incremented immediately and not checked as this leads to racey behavior
- increment+open/create needs to be atomic
- decrement+remove should also be atomic
- Can probably keep a reference to the fd and use it multiple times instead of opening multiple times

miaoyq replied:
>Can probably keep a reference to the fd and use it multiple times instead of opening multiple times

I think it may be hard to use the same fd multiple times, because the offset of logfile may be changed when  logfile is read each time, this will cause confusion if a logfile is read by multiple readers at the same time.

cpuguy83 replied:
You're right we're using seek here. The implementation could be updated to use `ReaderAt` to allow this, but could be done separately.

cpuguy83 commented on daemon/logger/jsonfilelog/jsonfilelog.go:
```diff
@@ -61,6 +61,17 @@ func New(info logger.Info) (logger.Logger, error) {
 	}
 
+	var compress = false
+	if maxFiles > 1 {
```
Maybe we should always check the compression option and error out if maxFiles is 0 or 1.

miaoyq replied:
This was addressing https://github.com/moby/moby/pull/29932#discussion_r149199872.
Log rotation is enabled only if `maxFiles > 1`, do we need to validate this when log rotation is disabled?

cpuguy83 reviewed (changes requested):
Looking good. The last case to deal with is parallel read requests.

miaoyq commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -48,10 +63,13 @@ func NewLogFile(logPath string, capacity int64, maxFiles int, marshaller logger.
 		currentSize:   size,
 		maxFiles:      maxFiles,
+		compress:      compress,
+		lastTimestamp: time.Time{},
```
Will do.

miaoyq commented:
@cpuguy83  Have addressed above comments, PTAL, Tks.

miaoyq commented:
@cpuguy83  I have tested the feature of parallel read requests.

Create a container like:
```
docker run -d --log-opt max-size=1m --log-opt max-file=5 --log-opt compression=gzip busybox /bin/sh -c 'while true; do sleep 0.001; echo "$(date)"; done'
```
Open 5 terminals and execute following command in each terminal to test parallel read requests:
```
watch -n 1 'docker logs --since 2000-01-01T00:00:00' containerId |wc  -l
```
It works well.

cpuguy83 commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -168,6 +254,19 @@ func (w *LogFile) ReadLogs(config logger.ReadConfig, watcher *logger.LogWatcher)
 			f.Close()
+			fileName := f.Name()
+			if strings.HasSuffix(fileName, tmpLogfileSuffix) {
+				w.filesRefCounter.Mu.Lock()
```
Can we do all the locking, counter changes in the refcounter object?

miaoyq replied:
Now Refcounter only count the tmp log files, so I think It's better to keep a single responsibility. :-)

cpuguy83 replied:
Right but we can do something like `counter.Reference(fileName)` and `counter.Dereference(filename)` rather than dealing with the internals of it here.

miaoyq replied:
Hum, I seemly hadn't got what you meant clearly :P, will do. Thanks! :-)

miaoyq replied:
@cpuguy83 Done, Please take another look, tks.

guangxuli commented:
/cc @guangxuli

miaoyq commented:
@cpuguy83 This PR has been created for a long time, could you please take another look, and merge it if it's OK?  Thanks so much.
I wish it can be merged this year. :-)

cpuguy83 commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -2,41 +2,107 @@ package loggerutils
+}
+
+// Reference increase the reference counter for specified logfile
+func (rc *RefCounter) Reference(fileName string, openRefFile func(fileName string) (*os.File, error)) (*os.File, error) {
```
Maybe we can change this `openRefFile` to `func(fileName string, exists bool) (*os.File, error)`
Then the implementation here would be:
```go
_, ok := rc.counter[fileName]
f, err := open(fileName, ok)
```

It may make sense to also change the fn name to `GetReference` in that case.

cpuguy83 commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -232,6 +391,45 @@ func (w *LogFile) openRotatedFiles() (files []*os.File, err error) {
+	_, err = pools.Copy(rs, rc)
+	if err != nil {
+		rs.Close()
+		os.Remove(rs.Name())
```
log remove error

cpuguy83 commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -224,6 +364,25 @@ func (w *LogFile) openRotatedFiles() (files []*os.File, err error) {
 				return nil, err
 			}
+
+			fileName := fmt.Sprintf("%s.%d.gz", w.f.Name(), i-1)
```
`w.f.Name()` may be racey here. We need to grab the rotation lock at the top of this function, I think.

cpuguy83 commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -93,48 +163,108 @@ func (w *LogFile) checkCapacityAndRotate() error {
+	compressWriter.Header.Extra, err = json.Marshal(&extra)
+	if err != nil {
+		logrus.Errorf("Failed to marshal JSON: %v", err)
+		return
```
Do we need to return early for this error since it's just an optimization?

miaoyq replied:
Will remve the return. :)

cpuguy83 commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -93,48 +163,108 @@ func (w *LogFile) checkCapacityAndRotate() error {
+	}
+	defer func() {
+		file.Close()
+		os.Remove(fileName)
```
log

cpuguy83 commented on daemon/logger/jsonfilelog/jsonfilelog.go:
```diff
@@ -127,6 +138,7 @@ func ValidateLogOpt(cfg map[string]string) error {
 		switch key {
 		case "max-file":
 		case "max-size":
+		case "compression":
```
It seems like we need some extra validation on this option.
Namely that it doesn't make sense if max-size is not set or if max-file is 0 or 1.

miaoyq replied:
Will do this, but I think we should valid this in 
 `func New(info logger.Info) (logger.Logger, error) `, `ValidateLogOpt` seems to valid the option name only.

miaoyq commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -93,48 +163,108 @@ func (w *LogFile) checkCapacityAndRotate() error {
+		return
+	}
+
+	file.Seek(0, os.SEEK_SET)
```
Will remove this.

miaoyq commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -93,48 +163,108 @@ func (w *LogFile) checkCapacityAndRotate() error {
+	file.Seek(0, os.SEEK_SET)
+	_, err = pools.Copy(compressWriter, file)
+	if err != nil {
+		logrus.Errorf("Error reading data to gzip log file: %v", err)
```
> Do we have a unit test for this failure scenario?

Still haven't test cases for this, will add this later

miaoyq commented:
@cpuguy83  Sorry for my delay, Thanks so much for your review again and again, :-)
 I have addressed all you comments, PTAL, Thanks!

AkihiroSuda commented:
@cpuguy83 PTAL 🙏

cpuguy83 commented:
Seeing a problem with readers cleaning up tmp files.
Basically as long as a reader is active the tmp files stay around, even when log files are rotated.
Can we release these files as soon as the reader has consumed their content?

Edge case I was playing with is having an active `docker logs -f` with rotated (and compressed files), let a (or some) rotations occur while the log reader is active, then start a new (2nd active) log reader.
The 2nd reader will get the cached tmp files even though they are now out of date.

Beyond the edge case, keeping these files around can lead to excessive disk usage for anyone who stays attached to the log.

cpuguy83 commented:
Not sure why this is randomly getting an EOF.

cpuguy83 commented:
```
DEBU[2018-02-14T16:12:36.180474800Z] begin logs                                    container=test method="(*Daemon).ContainerLogs" module=daemon

goroutine 203 [running]:
runtime.throw(0x1e04846, 0x21)
	/usr/local/go/src/runtime/panic.go:605 +0x95 fp=0xc420133db8 sp=0xc420133d98 pc=0x42cd95
sync.throw(0x1e04846, 0x21)
	/usr/local/go/src/runtime/panic.go:594 +0x35 fp=0xc420133dd8 sp=0xc420133db8 pc=0x42ccf5
sync.(*RWMutex).RUnlock(0xc4209b25b0)
	/usr/local/go/src/sync/rwmutex.go:71 +0x50 fp=0xc420133df8 sp=0xc420133dd8 pc=0x467a90
github.com/docker/docker/daemon/logger/loggerutils.(*LogFile).ReadLogs(0xc4209b25a0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xffffffffffffffff, 0x1, 0xc42100a9c0)
	/go/src/github.com/docker/docker/daemon/logger/loggerutils/logfile.go:298 +0xec fp=0xc420133f30 sp=0xc420133df8 pc=0xdf042c
github.com/docker/docker/daemon/logger/jsonfilelog.(*JSONFileLogger).readLogs(0xc420c0c8a0, 0xc42100a9c0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xffffffffffffffff, 0x1)
	/go/src/github.com/docker/docker/daemon/logger/jsonfilelog/read.go:30 +0xd4 fp=0xc420133f90 sp=0xc420133f30 pc=0xdf63c4
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2337 +0x1 fp=0xc420133f98 sp=0xc420133f90 pc=0x45e051
created by github.com/docker/docker/daemon/logger/jsonfilelog.(*JSONFileLogger).ReadLogs
	/go/src/github.com/docker/docker/daemon/logger/jsonfilelog/read.go:19 +0x13e
```

miaoyq commented:
@cpuguy83  Sorry for my delay due to Chinese New Year holidays, I have fixed this panic.

cpuguy83 reviewed (approved):
LGTM

Needs a minor rebase.

Commit:
add compress option for 'jsonfiles' log driver
This PR adds support for compressibility of log file.
I added a new option conpression for the jsonfile log driver,
this option allows the user to specify compression algorithm to
compress the log files. By default, the log files will be
not compressed. At present, only support 'gzip'.

Signed-off-by: Yanqiang Miao <miao.yanqiang@zte.com.cn>

'docker logs' can read from compressed files

Signed-off-by: Yanqiang Miao <miao.yanqiang@zte.com.cn>

Add Metadata to the gzip header, optmize 'readlog'

Signed-off-by: Yanqiang Miao <miao.yanqiang@zte.com.cn>

miaoyq commented:
@cpuguy83 Thanks a lot, rebased.

thaJeztah commented:
to other maintainers; having a quick look at this one; don't merge yet 😅

thaJeztah commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -2,42 +2,104 @@ package loggerutils // import "github.com/docker/docker/daemon/logger/loggerutil
-	marshal       logger.MarshalFunc
-	createDecoder makeDecoderFunc
-	perms         os.FileMode
+	logPath         string
```
Actually, given that `LogFile.LogPath()` seems to be unused, it looks like it's only used in `checkCapacityAndRotate()`. Can't we just use `f.Name()`?

Something like:

```patch
diff --git a/daemon/logger/loggerutils/logfile.go b/daemon/logger/loggerutils/logfile.go
index e0f21f46c..428f4786b 100644
--- a/daemon/logger/loggerutils/logfile.go
+++ b/daemon/logger/loggerutils/logfile.go
@@ -79,7 +79,6 @@ func (rc *refCounter) Dereference(fileName string) error {
 
 // LogFile is Logger implementation for default Docker logging.
 type LogFile struct {
-	logPath         string
 	mu              sync.RWMutex // protects the logfile access
 	f               *os.File     // store for closing
 	closed          bool
@@ -111,7 +110,6 @@ func NewLogFile(logPath string, capacity int64, maxFiles int, compress bool, mar
 	}
 
 	return &LogFile{
-		logPath:         logPath,
 		f:               log,
 		capacity:        capacity,
 		currentSize:     size,
@@ -162,15 +160,16 @@ func (w *LogFile) checkCapacityAndRotate() error {
 
 	if w.currentSize >= w.capacity {
 		w.rotateMu.Lock()
+		fname := w.f.Name()
 		if err := w.f.Close(); err != nil {
 			w.rotateMu.Unlock()
 			return errors.Wrap(err, "error closing file")
 		}
-		if err := rotate(w.logPath, w.maxFiles, w.compress); err != nil {
+		if err := rotate(fname, w.maxFiles, w.compress); err != nil {
 			w.rotateMu.Unlock()
 			return err
 		}
-		file, err := os.OpenFile(w.logPath, os.O_WRONLY|os.O_TRUNC|os.O_CREATE, w.perms)
+		file, err := os.OpenFile(fname, os.O_WRONLY|os.O_TRUNC|os.O_CREATE, w.perms)
 		if err != nil {
 			w.rotateMu.Unlock()
 			return err
@@ -185,7 +184,7 @@ func (w *LogFile) checkCapacityAndRotate() error {
 		}
 
 		go func() {
-			compressFile(w.logPath+".1", w.lastTimestamp)
+			compressFile(fname+".1", w.lastTimestamp)
 			w.rotateMu.Unlock()
 		}()
 	}
@@ -262,11 +261,6 @@ func compressFile(fileName string, lastTimestamp time.Time) {
 	}
 }
 
-// LogPath returns the location the given writer logs to.
-func (w *LogFile) LogPath() string {
-	return w.logPath
-}
-
 // MaxFiles return maximum number of files
 func (w *LogFile) MaxFiles() int {
 	return w.maxFiles
```


Or possibly even:

```patch
diff --git a/daemon/logger/loggerutils/logfile.go b/daemon/logger/loggerutils/logfile.go
index e0f21f46c..5ad023423 100644
--- a/daemon/logger/loggerutils/logfile.go
+++ b/daemon/logger/loggerutils/logfile.go
@@ -79,7 +79,6 @@ func (rc *refCounter) Dereference(fileName string) error {
 
 // LogFile is Logger implementation for default Docker logging.
 type LogFile struct {
-	logPath         string
 	mu              sync.RWMutex // protects the logfile access
 	f               *os.File     // store for closing
 	closed          bool
@@ -111,7 +110,6 @@ func NewLogFile(logPath string, capacity int64, maxFiles int, compress bool, mar
 	}
 
 	return &LogFile{
-		logPath:         logPath,
 		f:               log,
 		capacity:        capacity,
 		currentSize:     size,
@@ -166,11 +164,11 @@ func (w *LogFile) checkCapacityAndRotate() error {
 			w.rotateMu.Unlock()
 			return errors.Wrap(err, "error closing file")
 		}
-		if err := rotate(w.logPath, w.maxFiles, w.compress); err != nil {
+		if err := rotate(w.f.Name(), w.maxFiles, w.compress); err != nil {
 			w.rotateMu.Unlock()
 			return err
 		}
-		file, err := os.OpenFile(w.logPath, os.O_WRONLY|os.O_TRUNC|os.O_CREATE, w.perms)
+		file, err := os.OpenFile(w.f.Name(), os.O_WRONLY|os.O_TRUNC|os.O_CREATE, w.perms)
 		if err != nil {
 			w.rotateMu.Unlock()
 			return err
@@ -185,7 +183,7 @@ func (w *LogFile) checkCapacityAndRotate() error {
 		}
 
 		go func() {
-			compressFile(w.logPath+".1", w.lastTimestamp)
+			compressFile(w.f.Name()+".1", w.lastTimestamp)
 			w.rotateMu.Unlock()
 		}()
 	}
@@ -262,11 +260,6 @@ func compressFile(fileName string, lastTimestamp time.Time) {
 	}
 }
 
-// LogPath returns the location the given writer logs to.
-func (w *LogFile) LogPath() string {
-	return w.logPath
-}
-
 // MaxFiles return maximum number of files
 func (w *LogFile) MaxFiles() int {
 	return w.maxFiles
```

cpuguy83 replied:
It's nice to store it because `w.f.Name()` can be racey  (since `w.f` will change) unless locking is involved meanwhile the actual path never changes.

thaJeztah replied:
Yes, I was looking at that, seems the only place `LogFile.f` is written to is during construction in  [`NewLogFile()`](https://github.com/moby/moby/blob/241c904e6f5fff020890a7641558e83a209c0bbd/daemon/logger/loggerutils/logfile.go#L52), and in [`checkCapacityAndRotate()`](https://github.com/moby/moby/blob/241c904e6f5fff020890a7641558e83a209c0bbd/daemon/logger/loggerutils/logfile.go#L109), and the filename never actually changes (it's always `/var/lib/docker/containers/<container-id>/<container-id>-json.log`).

Perhaps the first option, using an intermediate variable (`fname := w.f.Name()`) would work then.

miaoyq replied:
I think what @thaJeztah said makes sense. @cpuguy83 WDYT?

cpuguy83 replied:
This works, but still need to lock to get the file name.

miaoyq replied:
@cpuguy83  Yeah, we can get the file name after `w.rotateMu.Lock()`.

thaJeztah commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -95,43 +161,112 @@ func (w *LogFile) checkCapacityAndRotate() error {
+		}
+
+		go func() {
+			compressFile(w.logPath+".1", w.lastTimestamp)
```
So, only time this could be a problem is if the next rotation takes place before compression is completed; not sure if that's a real issue, so just thinking out loud. I did some extreme test;

```patch
diff --git a/daemon/logger/loggerutils/logfile.go b/daemon/logger/loggerutils/logfile.go
index e0f21f46c..40bbba485 100644
--- a/daemon/logger/loggerutils/logfile.go
+++ b/daemon/logger/loggerutils/logfile.go
@@ -254,7 +254,7 @@ func compressFile(fileName string, lastTimestamp time.Time) {
                // Here log the error only and don't return since this is just an optimization.
                logrus.Warningf("Failed to marshal JSON: %v", err)
        }
-
+       time.Sleep(50 * time.Second)
        _, err = pools.Copy(compressWriter, file)
        if err != nil {
                logrus.WithError(err).WithField("module", "container.logs").WithField("file", fileName).Error("Error compressing log file")
```

Running with that, rotation is waiting for compression to complete, and killing the container became problematic, which is likely expected, just thinking if there would be real-world situations where that would happen.

cpuguy83 replied:
Yeah, this should be expected. There's possibly more we can do to prevent blocking on killing the container, but there's already a lot going on with locking for this case.

thaJeztah replied:
Yup, it was probably a bit extreme, just had my "QA hat" on, and seeing what possible things could be problematic. Don't think it should be a blocker for this feature.

thaJeztah reviewed (changes requested):
Left a comment about naming of the option (`compression` -> `compress`), and some thoughs

cpuguy83 commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -95,43 +161,112 @@ func (w *LogFile) checkCapacityAndRotate() error {
+}
+
+// LogPath returns the location the given writer logs to.
+func (w *LogFile) LogPath() string {
```
Yeah, I guess this isn't used anymore.

miaoyq replied:
Will remove this.

miaoyq commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -95,43 +161,112 @@ func (w *LogFile) checkCapacityAndRotate() error {
 		return nil
 	}
+
+	extension := ""
```
Will do.

thaJeztah commented:
ping @cpuguy83 could you look at https://github.com/moby/moby/pull/29932#discussion_r174340541 if you're ok with changing it to that?

miaoyq commented:
@cpuguy83 @thaJeztah Updated.

cpuguy83 reviewed (approved):
LGTM

thaJeztah reviewed (approved):
LGTM!!

thanks for this, I know it took a while 😅

thaJeztah commented:
hm, looks like this is a flaky test on z;

```
19:26:10 ----------------------------------------------------------------------
19:26:10 FAIL: docker_api_attach_test.go:98: DockerSuite.TestPostContainersAttach
19:26:10 
19:26:10 docker_api_attach_test.go:211:
19:26:10     c.Assert(actualStdout.Bytes(), checker.DeepEquals, []byte("hello\nsuccess"), check.Commentf("Attach didn't return the expected data from stdout"))
19:26:10 ... obtained []uint8 = []byte{0x73, 0x75, 0x63, 0x63, 0x65, 0x73, 0x73}
19:26:10 ... expected []uint8 = []byte{0x68, 0x65, 0x6c, 0x6c, 0x6f, 0xa, 0x73, 0x75, 0x63, 0x63, 0x65, 0x73, 0x73}
19:26:10 ... Attach didn't return the expected data from stdout
19:26:10 
```

miaoyq commented:
Thanks @cpuguy83 @thaJeztah  :smile:

thaJeztah commented:
argh, and now it failed on another flaky test (https://github.com/moby/moby/pull/36551 / https://github.com/moby/moby/issues/36547)
